### ISSUE TITLE
Compute mAP at different IoU

### ISSUE THREAD

**Issue by toobatehreem** [ultralytics/ultralytics#17815]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

The current validation outputs mAP50, 70 and 50:95, how can I compute mAP25 or mAP at any other IoU?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @toobatehreem, thank you for your question and for your interest in Ultralytics 🚀! This is an automated response to guide you while an Ultralytics engineer also reviews your inquiry and will assist you shortly.

We recommend visiting the [Docs](https://docs.ultralytics.com) as many common questions are addressed there, along with detailed examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it effectively.

Since your query appears to be about custom evaluation metrics ❓, including mAP computation at specific IoUs, we suggest providing as much information as possible to clarify your use case, such as your validation process details, configurations, or any relevant code snippets.

## Community Support

Get involved with our vibrant Ultralytics community for additional insights:
- For real-time chat, join [Discord](https://ultralytics.com/discord) 🎧
- Participate in in-depth discussions on [Discourse](https://community.ultralytics.com)
- Engage with others on our [Subreddit](https://reddit.com/r/ultralytics)

## Upgrade

First, ensure you're using the latest `ultralytics` version in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/), as updates may include relevant improvements or features for your use case. Upgrade with the following command:

```bash
pip install -U ultralytics
```

## Environments

For your experiments, consider verified computing environments with preinstalled dependencies:
- **Notebooks**: Free GPU-powered options:
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **AWS**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Utilize our [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Check our repository's CI status here:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If the badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, verifying functionality across platforms.

Looking forward to seeing how we can assist! 🚀

**Comment by Y-T-G** [comment]
You can't calculate IoU at 25 because that's part of the metrics. You can only find IoU between 50 to 95 in steps of 5.

```python
>>> results = model.val(data="coco128.yaml")
>>> results.box.all_ap[:, 0].mean() #mAP50
0.6741076875166357
>>> results.box.all_ap[:, 1].mean() #mAP55
0.6592387934367356
```

0 is for IoU 50, 1 is 55 and so on until 9 which is for 95.

**Comment by toobatehreem** [comment]
> You can't calculate IoU at 25 because that's not part of the metrics. You can only find IoU between 50 to 95 in steps of 5.
> 
> ```python
> >>> results = model.val(data="coco128.yaml")
> >>> results.box.all_ap[:, 0].mean() #mAP50
> 0.6741076875166357
> >>> results.box.all_ap[:, 1].mean() #mAP55
> 0.6592387934367356
> ```
> 
> 0 is for IoU 50, 1 is 55 and so on until 9 which is for 95.

what about mAP25?

**Comment by toobatehreem** [comment]
> You can't calculate IoU at 25 because that's not part of the metrics. You can only find IoU between 50 to 95 in steps of 5.
> 
> ```python
> >>> results = model.val(data="coco128.yaml")
> >>> results.box.all_ap[:, 0].mean() #mAP50
> 0.6741076875166357
> >>> results.box.all_ap[:, 1].mean() #mAP55
> 0.6592387934367356
> ```
> 
> 0 is for IoU 50, 1 is 55 and so on until 9 which is for 95.

is there any other way to compute mAP at IoU 25? The mAP at 50 IoU isn't good enough so I want to try some leniency with IoU thresholding.

**Comment by Y-T-G** [comment]
You will have to edit the source code to add those IoUs

**Comment by toobatehreem** [comment]
> You will have to edit the source code to add those IoUs

Okay, I have found it, it's inside yolo/val.py

**Comment by glenn-jocher** [comment]
That's correct! The IoU thresholds are defined in `yolo/val.py`, specifically in the section where performance metrics like mAP are calculated. To include mAP at IoU 25, you'd need to modify the IoU list used for metric computation. Be cautious and ensure thorough testing after making such changes. If you're unsure, consider submitting a pull request for review.

**Comment by xmj-7** [comment]
> > You will have to edit the source code to add those IoUs
> 
> Okay, I have found it, it's inside yolo/val.py

@toobatehreem bro，have you successfully converted map50 to map25？

**Comment by toobatehreem** [comment]
> > > You will have to edit the source code to add those IoUs
> > 
> > 
> > Okay, I have found it, it's inside yolo/val.py
> 
> @toobatehreem bro，have you successfully converted map50 to map25？
yes, change line 41 in yolo/detect/val.py to self.iouv = torch.linspace(0, 0.95, 20)

then you can access mAP25 as: results.box.all_ap[:, 5].mean()
mAP50 as: results.box.all_ap[:, 10].mean()
mAP50:95 as: results.box.all_ap[:, 10:].mean()

**Comment by glenn-jocher** [comment]
Thank you for sharing the solution! Modifying `yolo/val.py` as you suggested is a great approach for customizing IoU thresholds. Just remember to thoroughly revalidate your results to ensure accuracy after implementing these changes. If further assistance is needed, feel free to ask!

**Comment by xmj-7** [comment]
> > > > You will have to edit the source code to add those IoUs
> > > 
> > > 
> > > Okay, I have found it, it's inside yolo/val.py
> > 
> > 
> > @toobatehreem bro，have you successfully converted map50 to map25？
> 
> yes, change line 41 in yolo/detect/val.py to self.iouv = torch.linspace(0, 0.95, 20)
> 
> then you can access mAP25 as: results.box.all_ap[:, 5].mean() mAP50 as: results.box.all_ap[:, 10].mean() mAP50:95 as: results.box.all_ap[:, 10:].mean()

thanks you so much, it works.
